### PR TITLE
Add emulator jacobian plots

### LINKED ISSUE #1560: Jacobian scoring into train_microphysics.
#1559

### PR DESCRIPTION
**Pull request by frodre**
In creating the fixed RNN, it was useful to use output sensitivity plots to investigate model behavior.  This PR adds jacobian plots to the WandB training process.

[Example](https://wandb.ai/ai2cm/crapcrap/runs/1s3wu2tq?workspace=user-aperkins)

Resolves #1560 

The jacobians are added to the end of the training script because reloading saved models had some missing metadata that made the gradient calculation much more difficult (probably custom layer data not saved in the trace).  Everything important is stored before attempting the gradient calculation.

Breaking Changes:
- Updates `TrainConfig` defaults to new datasets and associated derived mapping hardcodes for tendencies,  (`_output` suffix -> `_after_precpd` suffix) 

Minor Adjustments:
- fixed the ability to specify tendency output names via config
- fixed some inconsistencies in the keras model outputs vs. the `MicrophysicsConfig.output_variables`

### CODE REVIEW

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 1]
```diff
@@ -27,31 +27,31 @@ def _register_derived_variables(mapping: DerivedMapping, timestep):
     # the data or configuration.
     @mapping.register(
         "tendency_of_air_temperature_due_to_microphysics",
-        required_inputs=["air_temperature_input", "air_temperature_output"],
+        required_inputs=["air_temperature_input", "air_temperature_after_precpd"],
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/data/transforms.py:30]: These are very brittle hardcodes.  Should probably switch to `Field` definition used by the conservative model and we can build up a derived mapping necessary for sample processing based on that.

File: external/fv3fit/fv3fit/emulation/layers/fields.py [hunk 2]
```diff
@@ -136,6 +136,7 @@ def __init__(
         sample_out: Optional[tf.Tensor] = None,
         denormalize: Optional[str] = None,
         enforce_positive: bool = False,
+        tendency_name: str = None,
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/layers/fields.py:139]: This is not a valid type hint. Type hints are pretty useful...I'm wondering if this part of the code is not being checked.
```suggestion
        tendency_name: Optional[str] = None,
```

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 3]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
             self.cache = False
 
 
+def get_model_output_sensitivities(
+    model: tf.keras.Model, sample: Mapping[str, tf.Tensor],
+):
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:213]: Output type hint please. mypy does nothing if there is no output type and it is nice to know at a glance what this function returns since the docstring isn't clear and the code is long.

This function does a lot...I'd consider separating the "jacobian of a dict" part from the "compute mean and normalize output part".

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 4]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
             self.cache = False
 
 
+def get_model_output_sensitivities(
+    model: tf.keras.Model, sample: Mapping[str, tf.Tensor],
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:212]: this function should work for any tensorflow code...so best to pose broadly.
```suggestion
    model: Callable[[Mapping[str, tf.Tensor]], Mapping[str, tf.Tensor]], sample: Mapping[str, tf.Tensor],
```

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 5]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
...
+        normalize_factors[name] = factor
+
+    input_data = {name: avg_profiles[name] for name in model.input_names}
+
+    with tf.GradientTape(persistent=True) as g:
+        g.watch(input_data)
+        outputs = model(input_data)
+
+    all_jacobians = {}
+    for out_name in model.output_names:
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:239]: this will remove the dependence on the `model.output_names` attribute and allow this to work with any tf function code and be more robust.

```suggestion
    for out_name in outputs:
```

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 6]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
             self.cache = False
 
 
+def get_model_output_sensitivities(
+    model: tf.keras.Model, sample: Mapping[str, tf.Tensor],
+):
+    """
+    Generate sensitivity jacabion for each output relative to a mean
+    value for each input
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:216]: Please describe outputs and the normalization strategy (e.g. what the units of the outputs)?

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 7]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
...
+
+        all_jacobians[out_name] = normalized
+
+    return all_jacobians
+
+
+OutputSensitivity = Mapping[str, np.ndarray]
+
+
+def plot_single_output_sensitivities(name: str, jacobians: OutputSensitivity):
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:256]: output type please.

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 8]
```diff
@@ -204,6 +208,93 @@ def __post_init__(self) -> None:
             self.cache = False
 
 
+def get_model_output_sensitivities(
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:211]: this file is getting rather long. Can you move the jacobian code (no plots) to a seperate module...and the plotting code elsewhere too...maybe there is an appropriate place already.

**Review by nbren12 — CHANGES_REQUESTED** [review]
Thanks for adding this...it would be nice if the plots had labels and units...some other nits below.

> The jacobians are added to the end of the training script because reloading saved models had some missing metadata that made the gradient calculation much more difficult 

Maybe add an xfail test demonstrating the problem so we can fix it later?

File: external/fv3fit/fv3fit/train_microphysics.py (hunk 3 above)
Comment (frodre) [reply external/fv3fit/fv3fit/train_microphysics.py:213]: Separated the jacobian  calculation from the jacobian normalizing.   Added the specifics for the functions to `emulation/keras.py` and the plotting to `wandb.py`.

File: external/fv3fit/tests/emulation/models/test_microphysics_config.py [hunk 9]
```diff
@@ -185,3 +186,37 @@ def test_RNN_downward_dependence():
             sensitivity = jacobian[output_level, input_level]
             if output_level > input_level and sensitivity != 0:
                 raise ValueError("Downwards dependence violated")
+
+
+@pytest.mark.xfail
+def test_saved_model_jacobian():
```
Comment (frodre) [inline external/fv3fit/tests/emulation/models/test_microphysics_config.py:192]: Failing test for jacobian calculation on saved RNNs

File: external/fv3fit/fv3fit/train_microphysics.py (hunk 8 above)
Comment (frodre) [reply external/fv3fit/fv3fit/train_microphysics.py]: Moved jacobian specifics into `emulation/keras.py` and the plotting into `wandb.py`.

**Comment by frodre** [comment]
New example with some axes labels: https://wandb.ai/ai2cm/crapcrap/runs/44jadbkr?workspace=user-aperkins

File: external/fv3fit/fv3fit/emulation/keras.py [hunk 10]
```diff
@@ -1,10 +1,11 @@
 import dataclasses
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/keras.py]: Maybe move to a new module named `jacobian.py` instead. `keras.py` is a pretty random assortment of things.

File: external/fv3fit/tests/emulation/test_keras.py [hunk 11]
```diff
@@ -171,3 +172,22 @@ def test_StandardLoss_prepare_arbitrary_kwargs():
...
+def test_get_model_output_sensitivities():
+    sample = {
+        "a": tf.random.normal((100, 5)),
+        "b": tf.random.normal((100, 5)),
+    }
+    sample["field"] = sample["a"] + sample["b"]
+
+    inputs = {"a": tf.keras.Input(5), "b": tf.keras.Input(5)}
+    out = tf.keras.layers.Lambda(lambda x: {"field": x["a"] + x["b"]})(inputs)
+    model = tf.keras.Model(inputs=inputs, outputs=out)
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/test_keras.py:132]: Optional...but does it need to be a model? I thought this would work with a plain old function. This would also ensure `get_model_output_sensitivies` is not using any of the more arcane APIs of `tf.keras.Model`.
```
def model(x):
    return {"field": x["a"] + x["b"]}
```

File: external/fv3fit/tests/emulation/test_keras.py [hunk 12]
```diff
@@ -112,3 +114,26 @@ def test_model_score_no_outputs():
...
+
+    jacobians = standardize_jacobians(get_jacobians(model, profiles), sample,)
+
+    assert set(jacobians) == {"field"}
+    assert set(jacobians["field"]) == {"a", "b"}
+    for j in jacobians["field"].values():
+        assert j.shape == (5, 5)
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/test_keras.py:139]: Required. You should check the actual values:
```suggestion
        n = 5  # maybe move to the top of the test
        np.testing.assert_array_almost_equal(jacobians["field"]["a"], np.eye(n))            
        np.testing.assert_array_almost_equal(jacobians["field"]["b"], np.eye(n))

```

**Review by nbren12 — APPROVED** [review]
Thanks for the improvements. I have a couple final suggestions below.

File: external/fv3fit/tests/emulation/test_keras.py (hunk 11 above)
Comment (frodre) [reply external/fv3fit/tests/emulation/test_keras.py]: It did not need a model to work, so I switched to use this
